lib/site-inspector.rb

```diff
@@ -127,41 +190,447 @@ def redirect
     end
   end
 
-  def to_json
-    to_hash.to_json
+  def http
+    details = {
+      endpoints: endpoints
+    }
+
+    # convenient shorthand for the extensive statements to come
+    combos = details[:endpoints]
+
+    # A domain is "canonically" at www if:
+    #  * at least one of its www endpoints responds
+    #  * both root endpoints are either down or redirect *somewhere*
+    #  * either both root endpoints are down, *or* at least one
+    #    root endpoint redirect should immediately go to
+    #    an *internal* www endpoint
+    # This is meant to affirm situations like:
+    #   http:// -> https:// -> https://www
+    #   https:// -> http:// -> https://www
+    # and meant to avoid affirming situations like:
+    #   http:// -> http://non-www,
+    #   http://www -> http://non-www
+    # or like:
+    #   https:// -> 200, http:// -> http://www
+
+    www = !!(
+      (
+        combos[:https][:www][:up] or
+        combos[:http][:www][:up]
+      ) and (
+        (
+          combos[:https][:root][:redirect] or
+          !combos[:https][:root][:up] or
+          combos[:https][:root][:https_bad_name] or
+          !combos[:https][:root][:status].to_s.start_with?("2")
+        ) and (
+          combos[:http][:root][:redirect] or
+          !combos[:http][:root][:up] or
+          !combos[:http][:root][:status].to_s.start_with?("2")
+        )
+      ) and (
```

Comment:
This is super gross. Is there any way to make the logic here more explicit (even if more verbose)? Perhaps breaking out a bunch of named methods?